dolma-cccc/domain_counter.py

```diff
@@ -0,0 +1,239 @@
+                            text_length_counter[suburl] += text_length  # Increment based on text length
+
+                    line_progress.update(1)  # Update progress bar for each processed line
+                except json.JSONDecodeError:
+                    continue
+
+    gc.collect()  # Explicitly trigger garbage collection to free memory
+    return suburl_counter, text_length_counter
+
+
+def extract_suburls(url):
+    parsed_url = urlparse(url)
+    domain = parsed_url.netloc
+    path_parts = parsed_url.path.strip('/').split('/')
+
+    suburls = [domain, ]
+    current_path = domain
+
+    for part in path_parts[:-1]:
+        current_path = f"{current_path}/{part}"
+        suburls.append(current_path)
+
+    return suburls
+
+
+def merge_counters(counter_list):
+    merged_counter = defaultdict(int)
+    for counter in counter_list:
+        for suburl, count in counter.items():
+            merged_counter[suburl] += count
+    return merged_counter
+
+
+def process_all_files_in_parallel(root_folder):
+    all_files = []
+
+    for dirpath, _, filenames in os.walk(root_folder):
+        for filename in filenames:
+            if filename.endswith('.json.gz'):
+                all_files.append(os.path.join(dirpath, filename))
+
+    def signal_handler(sig, frame):
+        print("Interrupt received, terminating...")
+        executor.shutdown(wait=False)
+        sys.exit(1)
+
+    # Register the signal handler
+    signal.signal(signal.SIGINT, signal_handler)
+
+    with tqdm(total=len(all_files), desc="Total files", position=0) as file_progress:
+        with ProcessPoolExecutor(max_workers=40) as executor:  # Adjust max_workers based on your system
+            future_to_file = {executor.submit(process_file, file): file for file in all_files}
+
+            suburl_results = []
+            text_length_results = []
+
+            try:
+                for future in tqdm(as_completed(future_to_file), total=len(future_to_file), desc="Processing files"):
+                    try:
+                        suburl_counter, text_length_counter = future.result()
+                        suburl_results.append(suburl_counter)
+                        text_length_results.append(text_length_counter)
+                    except Exception as e:
+                        print(f"Error processing file {future_to_file[future]}: {e}")
+                    finally:
+                        file_progress.update(1)  # Update the outer file-level progress bar
+
+                wait(future_to_file.keys(), return_when=FIRST_EXCEPTION)
+
+            except KeyboardInterrupt:
+                print("Ctrl+C pressed, terminating...")
+                executor.shutdown(wait=False)
+                sys.exit(1)
+
+    final_suburl_counter = merge_counters(suburl_results)
+    final_text_length_counter = merge_counters(text_length_results)
+
+    sorted_suburls = sorted(final_suburl_counter.items(), key=lambda item: item[1], reverse=True)
```

Comment:
You're using a counter object, you can do `final_suburl_counter.most_common(10)` to get the top 10 sub urls by frequency.